File: stgit/commands/refresh.py
from stgit import argparse, utils
from stgit.argparse import opt
from stgit.commands.common import (
    CmdException,
    DirectoryHasRepository,
    check_head_top_equal,
    run_commit_msg_hook,
)
from stgit.config import config
from stgit.lib.edit import auto_edit_patch, interactive_edit_patch
from stgit.lib.git import CommitData, IndexAndWorktree
from stgit.lib.transaction import (
    StackTransaction,
    TransactionException,
    TransactionHalted,
)
from stgit.out import out
from stgit.run import RunException
from stgit.utils import get_hook

__copyright__ = """
Copyright (C) 2005, Catalin Marinas <catalin.marinas@gmail.com>
Copyright (C) 2008, Karl Hasselström <kha@treskal.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see http://www.gnu.org/licenses/.
"""

help = 'Generate a new commit for the current patch'
kind = 'patch'
usage = ['[options] [--] [<files or dirs>]']
description = """
Include the latest work tree and index changes in the current patch.
This command generates a new git commit object for the patch; the old
commit is no longer visible.

Refresh will warn if the index is dirty, and require use of either the
'--index' or '--force' options to override this check. This is to prevent
accidental full refresh when only some changes were staged using git add
interative mode.

You may optionally list one or more files or directories relative to
the current working directory; if you do, only matching files will be
updated.

Behind the scenes, stg refresh first creates a new temporary patch
with your updates, and then merges that patch into the patch you asked
to have refreshed. If you asked to refresh a patch other than the
topmost patch, there can be conflicts; in that case, the temporary
patch will be left for you to take care of, for example with stg
squash.

The creation of the temporary patch is recorded in a separate entry in
the patch stack log; this means that one undo step will undo the merge
between the other patch and the temp patch, and two undo steps will
additionally get rid of the temp patch.

Additionally, the '--spill' option resets the topmost patch, emptying
the patch while leaving the patch's changes intact in the worktree."""

args = ['dirty_files']
options = [
    opt(
        '-u',
        '--update',
        action='store_true',
        short='Only update the current patch files',
    ),
    opt(
        '-i',
        '--index',
        action='store_true',
        short='Refresh from index instead of worktree',
        long="""
        Instead of setting the patch top to the current contents of
        the worktree, set it to the current contents of the index.""",
    ),
    opt(
        '-F',
        '--force',
        action='store_true',
        short='Force refresh even if index is dirty',
        long="""
        Instead of warning the user when some work has already been staged
        (such as with git add interactive mode) force a full refresh.""",
    ),
    opt(
        '-p',
        '--patch',
        args=['other_applied_patches', 'unapplied_patches'],
        short='Refresh (applied) PATCH instead of the top patch',
    ),
    opt(
        '-e',
        '--edit',
        action='store_true',
        short='Invoke an editor for the patch description',
    ),
    opt(
        '-d',
        '--diff',
        action='store_true',
        short='Show diff when editing patch description',
    ),
    opt(
        '-a',
        '--annotate',
        metavar='NOTE',
        short='Annotate the patch log entry',
    ),
    opt(
        '-s',
        '--submodules',
        action='store_true',
        short='Include submodules when refreshing patch contents',
    ),
    opt(
        '--no-submodules',
        action='store_false',
        dest='submodules',
        short='Exclude submodules when refreshing patch contents',
    ),
    opt(
        '--spill',
        action='store_true',
        dest='spill',
        short='Spill patch content to worktree and index, erasing patch content.',
    ),
]
options.extend(argparse.message_options(save_template=False))
options.extend(argparse.hook_options())
options.extend(argparse.trailer_options())
options.extend(argparse.author_options())
options.extend(argparse.diff_opts_option())

directory = DirectoryHasRepository()


def get_patch(stack, given_patch):
    """Get the name of the patch we are to refresh."""
    if given_patch:
        patch_name = given_patch
        if patch_name not in stack.patches:
            raise CmdException('%s: no such patch' % patch_name)
        return patch_name
    else:
        if not stack.patchorder.applied:
            raise CmdException(
                'Cannot refresh top patch because no patches are applied'
            )
        return stack.patchorder.applied[-1]


def list_files(stack, patch_name, path_limits, update, submodules):
    """Figure out which files to update."""
    iw = stack.repository.default_iw
    paths = iw.changed_files(stack.head.data.tree, path_limits or [])
    if update:
        # --update: Restrict update to the paths that were already part of the patch.
        patch_paths = set()
        commit = stack.patches[patch_name]
        for dt in stack.repository.diff_tree_files(
            commit.data.parent.data.tree, commit.data.tree
        ):
            _, _, _, _, _, oldname, newname = dt
            patch_paths.add(oldname)
            patch_paths.add(newname)
        paths &= patch_paths
    else:
        # Avoid including submodule files by default. This is to ensure that
        # users in repositories with submodueles do not accidentally include
        # submodule changes to patches just because they happen to have not
        # run "git submodule update" prior to running stg refresh. We won't
        # exclude them if we're explicitly told to include them, or if we're
        # given explicit paths.
        if not path_limits and not submodules:
            paths -= stack.repository.submodules(stack.head.data.tree)
    return paths


def write_tree(stack, paths, use_temp_index):
    """Possibly update the index, and then write its tree.

    If any path limiting is in effect, use a temp index.

    :return: The written tree.
    :rtype: :class:`Tree<stgit.git.Tree>`

    """

    def go(index):
        if paths:
            iw = IndexAndWorktree(index, stack.repository.default_worktree)
            iw.update_index(paths)
        return index.write_tree()

    if use_temp_index:
        with stack.repository.temp_index() as index:
            try:
                index.read_tree(stack.head)
                return go(index)
            finally:
                stack.repository.default_iw.update_index(paths)
    else:
        return go(stack.repository.default_index)


def make_temp_patch(stack, patch_name, tree):
    """Commit tree to temp patch, in a complete transaction."""
    commit = stack.repository.commit(
        CommitData(
            tree=tree,
            parents=[stack.head],
            message='Refresh of %s' % patch_name,
        )
    )
    temp_name = stack.patches.make_name('refresh-temp')
    trans = StackTransaction(stack)
    trans.patches[temp_name] = commit
    trans.applied.append(temp_name)
    return (
        trans.run(
            'refresh (create temporary patch)',
            stack.repository.default_iw,
            print_current_patch=False,
        ),
        temp_name,
    )


def absorb_applied(trans, iw, patch_name, temp_name, edit_fun):
    """Absorb temp patch into the given patch, which must be applied.

    If the absorption succeeds, call ``edit_fun`` on the resulting
    :class:`stgit.lib.git.CommitData` before committing it, then commit the return
    value.

    :returns: True if the temp patch is successfully absorbed or False if there remain
              conflicts for the user to resolve

    """
    temp_absorbed = False
    try:
        # Pop any patch on top of the patch we're refreshing.
        to_pop = trans.applied[trans.applied.index(patch_name) + 1 :]
        if len(to_pop) > 1:
            popped_extra = trans.pop_patches(lambda pn: pn in to_pop)
            assert not popped_extra  # no other patches were popped
            trans.push_patch(temp_name, iw)
        top_name = to_pop.pop()
        assert top_name == temp_name

        # Absorb the temp patch.
        temp_cd = trans.patches[temp_name].data
        assert trans.patches[patch_name] == temp_cd.parent
        cd, new_patch_name = edit_fun(
            trans.patches[patch_name].data.set_tree(temp_cd.tree)
        )
        trans.patches[patch_name] = trans.stack.repository.commit(cd)
        if new_patch_name and patch_name != new_patch_name:
            trans.rename_patch(patch_name, new_patch_name)

        popped_extra = trans.delete_patches(lambda pn: pn == temp_name, quiet=True)
        assert not popped_extra  # the temp patch was topmost
        temp_absorbed = True

        # Push back any patch we were forced to pop earlier.
        for pn in to_pop:
            trans.push_patch(pn, iw)
    except TransactionHalted:
        pass
    return temp_absorbed


def absorb_unapplied(trans, iw, patch_name, temp_name, edit_fun):
    """Absorb the temp patch into the given patch, which must be unapplied.

    If the absorption succeeds, call ``edit_fun`` on the resulting
    :class:`stgit.lib.git.CommitData` before committing it, then commit the return
    value.

    :param iw: Not used.
    :returns: True if the temp patch is absorbed successfully or False
              if user resolvable conflicts remain

    """
    # Pop the temp patch.
    popped_extra = trans.pop_patches(lambda pn: pn == temp_name)
    assert not popped_extra  # the temp patch was topmost

    # Try to create the new tree of the refreshed patch. (This is the
    # same operation as pushing the temp patch onto the patch we're
    # trying to refresh -- but we don't have a worktree to spill
    # conflicts to, so if the simple merge doesn't succeed, we have to
    # give up.)
    patch_cd = trans.patches[patch_name].data
    temp_cd = trans.patches[temp_name].data
    new_tree = trans.stack.repository.simple_merge(
        base=temp_cd.parent.data.tree,
        ours=patch_cd.tree,
        theirs=temp_cd.tree,
    )
    if new_tree:
        # It worked. Refresh the patch with the new tree, and delete
        # the temp patch.
        cd, new_patch_name = edit_fun(patch_cd.set_tree(new_tree))
        trans.patches[patch_name] = trans.stack.repository.commit(cd)
        if new_patch_name and patch_name != new_patch_name:
            trans.rename_patch(patch_name, new_patch_name)
        popped_extra = trans.delete_patches(lambda pn: pn == temp_name, quiet=True)
        assert not popped_extra  # the temp patch was not applied
        return True
    else:
        # Nope, we couldn't create the new tree, so we'll just have to
        # leave the temp patch for the user.
        return False


def absorb(stack, patch_name, temp_name, edit_fun, annotate=None):
    """Absorb the temp patch into the target patch."""
    log_msg = 'refresh'
    if annotate:
        log_msg += '\n\n' + annotate

    check_head_top_equal(stack)
    trans = StackTransaction(stack)
    iw = stack.repository.default_iw
    if patch_name in trans.applied:
        absorb_func = absorb_applied
    else:
        absorb_func = absorb_unapplied

    if absorb_func(trans, iw, patch_name, temp_name, edit_fun):

        def info_msg():
            pass

    else:

        def info_msg():
            out.warn(
                'The new changes did not apply cleanly to %s.' % patch_name,
                'They were saved in %s.' % temp_name,
            )

    r = trans.run(log_msg, iw)
    info_msg()
    return r


def __refresh_spill(annotate):
    stack = directory.repository.current_stack

    # Fetch the topmost patch.
    patchname = get_patch(stack, None)

    cd = stack.patches[patchname].data

    # Set the tree of the patch to the parent.
    cd = cd.set_tree(cd.parent.data.tree)

    log_msg = 'refresh (spill)'
    if annotate:
        log_msg += '\n\n' + annotate

    check_head_top_equal(stack)
    trans = StackTransaction(stack, allow_conflicts=True)
    trans.patches[patchname] = stack.repository.commit(cd)
    try:
        # Either a complete success, or a conflict during push. But in
        # either case, we've successfully effected the edits the user
        # asked us for.
        return trans.run(log_msg)
    except TransactionException:
        # Transaction aborted -- we couldn't check out files due to
        # dirty index/worktree. The edits were not carried out.
        out.error('Unable to spill the topmost patch')
        return utils.STGIT_COMMAND_ERROR


def __refresh(
    args,
    force=False,
    target_patch=None,
    message=None,
    author=None,
    trailers=None,
    annotate=None,
    use_temp_index=False,
    refresh_from_index=False,
    only_update_patchfiles=False,
    include_submodules=False,
    no_verify=False,
    invoke_editor=False,
    edit_diff=False,
    diff_flags=(),
):
    stack = directory.repository.current_stack

    patch_name = get_patch(stack, target_patch)

    if refresh_from_index:
        paths = set()
    else:
        paths = list_files(
            stack,
            patch_name,
            args,
            only_update_patchfiles,
            include_submodules,
        )

    # Make sure there are no conflicts in the files we want to
    # refresh.
    if stack.repository.default_index.conflicts() & paths:
        raise CmdException('Cannot refresh -- resolve conflicts first')

    # Make sure the index is clean before performing a full refresh
    if not refresh_from_index and not force:
        if not (
            stack.repository.default_index.is_clean(stack.head)
            or stack.repository.default_iw.worktree_clean()
        ):
            raise CmdException(
                'The index is dirty. Did you mean --index? '
                'To force a full refresh use --force.'
            )

    check_head_top_equal(stack)

    # Update index and write tree
    tree = write_tree(stack, paths, use_temp_index=use_temp_index)

    # Run pre-commit hook, if fails, abort refresh
    if not no_verify:
        pre_commit_hook = get_hook(
            stack.repository,
            'pre-commit',
            extra_env={} if invoke_editor else {'GIT_EDITOR': ':'},
        )
        if pre_commit_hook:
            try:
                pre_commit_hook()
            except RunException:
                raise CmdException(
                    'pre-commit hook failed, review the changes using `stg diff`, '
                    'run `stg add` to add them to index and run `stg refresh` again'
                )
            else:
                # Update index and rewrite tree if hook updated files in index
                if not stack.repository.default_index.is_clean(tree):
                    tree = write_tree(stack, paths, use_temp_index=use_temp_index)

    # Commit tree to temp patch, and absorb it into the target patch.
    retval, temp_name = make_temp_patch(stack, patch_name, tree)

    if retval != utils.STGIT_SUCCESS:
        return retval

    def edit_fun(cd):
        orig_msg = cd.message
        new_msg = None
        if message is not None:
            new_msg = message.encode(config.get('i18n.commitencoding'))
        cd = auto_edit_patch(
            stack.repository,
            cd,
            msg=new_msg,
            author=author,
            trailers=trailers,
        )
        new_patch_name = None
        if invoke_editor:
            cd, new_patch_name, failed_diff = interactive_edit_patch(
                stack.repository, cd, patch_name, edit_diff, diff_flags
            )
            assert not failed_diff
        if not no_verify and (invoke_editor or cd.message != orig_msg):
            cd = run_commit_msg_hook(stack.repository, cd, invoke_editor)
        # Refresh the committer information
        return cd.set_committer(None), new_patch_name

    return absorb(stack, patch_name, temp_name, edit_fun, annotate=annotate)


def func(parser, options, args):
    """Generate a new commit for the current or given patch."""

    if options.spill:
        if len(args) > 0:
            # TODO: would be nice if path limiting could be used with spill.
            raise CmdException('Cannot use path limiting with --spill')
        for opt_name, opt_value in [
            ('--index', options.index),
            ('--edit', options.edit),
            ('--update', options.update),
            ('--patch', options.patch),
            ('--force', options.force),
            ('--no-verify', options.no_verify),
            ('--sign', options.trailers),
            ('--ack', options.trailers),
            ('--review', options.trailers),
        ]:
            if opt_value:
                raise CmdException('Cannot combine --spill with %s' % opt_name)
        return __refresh_spill(annotate=options.annotate)
    else:
        # Catch illegal argument combinations.
        is_path_limiting = bool(args or options.update)
        if options.index and is_path_limiting:
            raise CmdException('Only full refresh is available with the --index option')

        if options.index and options.force:
            raise CmdException(
                'You cannot --force a full refresh when using --index mode'
            )

        if options.update and options.submodules:
            raise CmdException(
                '--submodules is meaningless when only updating modified files'
            )

        if options.index and options.submodules:
            raise CmdException(
                '--submodules is meaningless when keeping the current index'
            )

        # If submodules was not specified on the command line, infer a default
        # from configuration.
        if options.submodules is None:
            options.submodules = config.getbool('stgit.refreshsubmodules')

        return __refresh(
            args,
            force=options.force,
            target_patch=options.patch,
            message=options.message,
            author=options.author,
            trailers=options.trailers,
            annotate=options.annotate,
            use_temp_index=is_path_limiting,
            refresh_from_index=options.index,
            only_update_patchfiles=options.update,
            include_submodules=options.submodules,
            no_verify=options.no_verify,
            invoke_editor=options.edit,
            edit_diff=options.diff,
            diff_flags=options.diff_flags,
        )

File: src/cmd/refresh.rs
// SPDX-License-Identifier: GPL-2.0-only

//! `stg refresh` implementation.

use std::{
    path::{Path, PathBuf},
    rc::Rc,
};

use anyhow::{anyhow, Result};
use clap::{Arg, ArgGroup, ArgMatches, ValueHint};
use indexmap::IndexSet;

use crate::{
    argset,
    color::get_color_stdout,
    ext::{CommitExtended, RepositoryExtended, SignatureExtended},
    hook::run_pre_commit_hook,
    patch::{patchedit, LocationConstraint, PatchLocator, PatchName},
    stack::{InitializationPolicy, Stack, StackAccess, StackStateAccess},
    stupid::{Status, StatusOptions, Statuses, Stupid, StupidContext},
    wrap::Message,
};

pub(super) const STGIT_COMMAND: super::StGitCommand = super::StGitCommand {
    name: "refresh",
    category: super::CommandCategory::PatchManipulation,
    make,
    run,
};

fn make() -> clap::Command {
    let app = clap::Command::new("refresh")
        .about("Incorporate worktree changes into current patch")
        .long_about(
            "Include the latest work tree and index changes in the \
             current patch. This command generates a new git commit \
             object for the patch; the old commit is no longer visible.\n\
             \n\
             Refresh will warn if the index is dirty, and require use of \
             either the '--index' or '--force' options to override this \
             check. This is to prevent accidental full refresh when only \
             some changes were staged using git add interactive mode.\n\
             \n\
             You may optionally list one or more files or directories \
             relative to the current working directory; if you do, only \
             matching files will be updated.\n\
             \n\
             Behind the scenes, stg refresh first creates a new \
             temporary patch with your updates, and then merges that \
             patch into the patch you asked to have refreshed. If you \
             asked to refresh a patch other than the topmost patch, \
             there can be conflicts; in that case, the temporary patch \
             will be left for you to take care of, for example with stg \
             squash.\n\
             \n\
             The creation of the temporary patch is recorded in a \
             separate entry in the patch stack log; this means that one \
             undo step will undo the merge between the other patch and \
             the temp patch, and two undo steps will additionally get \
             rid of the temp patch.",
        )
        .arg(
            Arg::new("pathspecs")
                .help("Only refresh files matching path")
                .value_name("path")
                .num_args(1..)
                .value_parser(clap::value_parser!(PathBuf)),
        )
        .next_help_heading("Refresh Options")
        .arg(
            Arg::new("update")
                .long("update")
                .short('u')
                .help("Only update the current patch files")
                .action(clap::ArgAction::SetTrue),
        )
        .arg(
            Arg::new("index")
                .long("index")
                .short('i')
                .help("Refresh from index instead of worktree")
                .long_help(
                    "Instead of setting the patch top to the current \
                     contents of the worktree, set it to the current \
                     contents of the index.",
                )
                .action(clap::ArgAction::SetTrue)
                .conflicts_with_all(["pathspecs", "update", "submodules", "force"]),
        )
        .arg(
            Arg::new("force")
                .long("force")
                .short('F')
                .help("Force refresh even if index is dirty")
                .long_help(
                    "Instead of warning the user when some work has \
                     already been staged (such as with git add \
                     interactive mode) force a full refresh.",
                )
                .action(clap::ArgAction::SetTrue),
        )
        .arg(argset::push_conflicts_arg())
        .arg(
            Arg::new("patch")
                .long("patch")
                .short('p')
                .help("Refresh (applied) <patch> instead of the top patch")
                .num_args(1)
                .value_name("patch")
                .value_hint(ValueHint::Other)
                .allow_hyphen_values(true)
                .value_parser(clap::value_parser!(PatchLocator)),
        )
        .arg(
            Arg::new("annotate")
                .long("annotate")
                .short('a')
                .help("Annotate the patch log entry with <note>")
                .num_args(1)
                .value_name("note")
                .value_hint(ValueHint::Other),
        )
        .arg(
            Arg::new("submodules")
                .long("submodules")
                .help("Include submodules in patch content")
                .action(clap::ArgAction::SetTrue)
                .conflicts_with("update"),
        )
        .arg(
            Arg::new("no-submodules")
                .long("no-submodules")
                .help("Exclude submodules in patch content")
                .action(clap::ArgAction::SetTrue),
        )
        .group(ArgGroup::new("submodule-group").args(["submodules", "no-submodules"]))
        .arg(
            Arg::new("spill")
                .long("spill")
                .help("OBSOLETE: use 'stg spill'")
                .hide(true)
                .action(clap::ArgAction::SetTrue),
        );

    patchedit::add_args(app, true, false)
}

fn run(matches: &ArgMatches) -> Result<()> {
    if matches.get_flag("spill") {
        return Err(anyhow!(
            "`stg refresh --spill` is obsolete; use `stg spill` instead"
        ));
    }

    let repo = gix::Repository::open()?;
    let stack = Stack::current(&repo, InitializationPolicy::AllowUninitialized)?;
    let config = repo.config_snapshot();
    let allow_push_conflicts = argset::resolve_allow_push_conflicts(&config, matches);

    stack.check_head_top_mismatch()?;

    let patchname = if let Some(patch_loc) = matches.get_one::<PatchLocator>("patch") {
        patch_loc
            .resolve_name(&stack)?
            .constrain(&stack, LocationConstraint::Visible)?
    } else if let Some(top_patchname) = stack.applied().last() {
        top_patchname.clone()
    } else {
        return Err(super::Error::NoAppliedPatches.into());
    };

    let tree_id = assemble_refresh_tree(
        &stack,
        matches,
        matches.get_flag("update").then_some(&patchname),
    )?;

    let mut log_msg = "refresh ".to_string();
    let opt_annotate = matches.get_one::<String>("annotate");

    // Make temp patch
    let temp_commit_id = stack.repo.commit_ex(
        &repo.get_author()?.override_author(matches),
        repo.get_committer()?,
        &Message::from(format!("Refresh of {patchname}")),
        tree_id,
        [stack.get_branch_head().id],
    )?;

    let temp_patchname = {
        let len_limit = None;
        let allow = vec![];
        let disallow: Vec<&PatchName> = stack.all_patches().collect();
        PatchName::make("refresh-temp", true, len_limit).uniquify(&allow, &disallow)
    };

    let stack = stack
        .setup_transaction()
        .with_output_stream(get_color_stdout(matches))
        .transact(|trans| trans.new_applied(&temp_patchname, temp_commit_id))
        .execute(&format!(
            "refresh {temp_patchname} (create temporary patch)"
        ))?;

    let mut absorb_success = false;
    stack
        .setup_transaction()
        .use_index_and_worktree(true)
        .with_output_stream(get_color_stdout(matches))
        .allow_push_conflicts(allow_push_conflicts)
        .transact(|trans| {
            if let Some(pos) = trans.applied().iter().position(|pn| pn == &patchname) {
                // Absorb temp patch into already applied patch
                let to_pop = trans.applied()[pos + 1..].to_vec();
                if to_pop.len() > 1 {
                    let popped_extra = trans.pop_patches(|pn| to_pop.contains(pn))?;
                    assert!(
                        popped_extra.is_empty(),
                        "only requested patches should be popped"
                    );
                    trans.push_patches(&[&temp_patchname], false)?;
                }

                let temp_commit = trans.get_patch_commit(&temp_patchname);

                let mut to_pop = to_pop;
                let top_name = to_pop.pop();
                assert_eq!(top_name.as_ref(), Some(&temp_patchname));

                let (new_patchname, new_commit_id) = match patchedit::EditBuilder::default()
                    .original_patchname(Some(&patchname))
                    .existing_patch_commit(trans.get_patch_commit(&patchname))
                    .override_tree_id(temp_commit.tree_id()?.detach())
                    .allow_diff_edit(false)
                    .allow_implicit_edit(false)
                    .allow_template_save(false)
                    .edit(trans, &repo, matches)?
                {
                    patchedit::EditOutcome::Edited {
                        new_patchname,
                        new_commit_id,
                    } => (new_patchname, new_commit_id),
                    patchedit::EditOutcome::TemplateSaved(_) => {
                        panic!("not allowed for refresh")
                    }
                };

                trans.delete_patches(|pn| pn == &temp_patchname)?;
                assert_eq!(Some(&patchname), trans.applied().last());
                if let Some(commit_id) = new_commit_id {
                    trans.update_patch(&patchname, commit_id)?;
                }
                if let Some(new_patchname) = new_patchname {
                    trans.rename_patch(&patchname, &new_patchname)?;
                } else {
                    log_msg.push_str(patchname.as_ref());
                }
                if let Some(annotation) = opt_annotate {
                    log_msg.push_str("\n\n");
                    log_msg.push_str(annotation);
                }

                trans.push_patches(&to_pop, false)?;
                absorb_success = true;
            } else {
                // Absorb temp patch into unapplied patch
                let popped_extra = trans.pop_patches(|pn| pn == &temp_patchname)?;
                assert!(popped_extra.is_empty());

                // Try to create the new tree of the refreshed patch.
                // This is the same as pushing the temp patch onto the target patch,
                // but without a worktree to spill conflicts to; so if the simple
                // merge fails, the refresh must be aborted.

                let patch_commit = trans.get_patch_commit(&patchname);
                let temp_commit = trans.get_patch_commit(&temp_patchname);
                let base = temp_commit.get_parent_commit()?.tree_id()?.detach();
                let ours = patch_commit.tree_id()?.detach();
                let theirs = temp_commit.tree_id()?.detach();

                if let Some(tree_id) = repo.stupid().with_temp_index(|stupid_temp| {
                    stupid_temp.read_tree(ours)?;
                    if stupid_temp.apply_treediff_to_index(base, theirs, true)? {
                        let tree_id = stupid_temp.write_tree()?;
                        Ok(Some(tree_id))
                    } else {
                        Ok(None)
                    }
                })? {
                    let (new_patchname, new_commit_id) = match patchedit::EditBuilder::default()
                        .original_patchname(Some(&patchname))
                        .existing_patch_commit(trans.get_patch_commit(&patchname))
                        .override_tree_id(tree_id)
                        .allow_diff_edit(false)
                        .allow_template_save(false)
                        .edit(trans, &repo, matches)?
                    {
                        patchedit::EditOutcome::Edited {
                            new_patchname,
                            new_commit_id,
                        } => (new_patchname, new_commit_id),
                        patchedit::EditOutcome::TemplateSaved(_) => {
                            panic!("not allowed for refresh")
                        }
                    };

                    if let Some(commit_id) = new_commit_id {
                        trans.update_patch(&patchname, commit_id)?;
                    }
                    if let Some(new_patchname) = new_patchname {
                        trans.rename_patch(&patchname, &new_patchname)?;
                        log_msg.push_str(new_patchname.as_ref());
                    } else {
                        log_msg.push_str(patchname.as_ref());
                    }
                    if let Some(annotation) = opt_annotate {
                        log_msg.push_str("\n\n");
                        log_msg.push_str(annotation);
                    }
                    trans.delete_patches(|pn| pn == &temp_patchname)?;
                    absorb_success = true;
                }
            }
            Ok(())
        })
        .execute(&log_msg)?;

    if !absorb_success {
        println!(
            "The new changes did not apply cleanly to {}. \
             They were saved in {}.",
            &patchname, &temp_patchname,
        );
    }

    Ok(())
}

fn determine_refresh_paths(
    stupid: &StupidContext,
    statuses: &Statuses,
    patch_commit: Option<&Rc<gix::Commit>>,
    force: bool,
) -> Result<IndexSet<PathBuf>> {
    let refresh_paths: IndexSet<&Path> = if let Some(patch_commit) = patch_commit {
        // Restrict update to the paths that were already part of the patch.
        let parent_tree_id = patch_commit.get_parent_commit()?.tree_id()?.detach();
        let diff_files =
            stupid.diff_tree_files(parent_tree_id, patch_commit.tree_id()?.detach())?;
        let patch_paths = diff_files.iter().collect::<IndexSet<_>>();

        statuses
            .iter()
            .filter_map(|entry| {
                let path = entry.path();
                if patch_paths.contains(path) {
                    Some(path)
                } else {
                    None
                }
            })
            .collect()
    } else {
        statuses.iter().map(|entry| entry.path()).collect()
    };

    // Ensure no conflicts in the files to be refreshed.
    statuses.check_conflicts_filter(|entry| refresh_paths.contains(entry.path()))?;

    // Ensure worktree and index states are valid for the given options.
    // Forcing means changes will be taken from both the index and worktree.
    // If not forcing, all changes must be either in the index or worktree,
    // but not both.
    if !force {
        let mut is_index_clean = true;
        let mut is_worktree_clean = true;
        for entry in statuses.iter() {
            if !matches!(entry.index_status(), Status::Unmodified) {
                is_index_clean = false;
            }
            if !matches!(entry.worktree_status(), Status::Unmodified) {
                is_worktree_clean = false;
            }
            if !is_index_clean && !is_worktree_clean {
                return Err(anyhow!(
                    "the index is dirty; consider using `--index` or `--force`",
                ));
            }
        }
    }

    // TODO: interrogate status once and avoid allocating PathBufs
    let refresh_paths = refresh_paths
        .iter()
        .map(|path| path.to_path_buf())
        .collect();
    Ok(refresh_paths)
}

fn write_tree(
    stack: &Stack,
    refresh_paths: &IndexSet<PathBuf>,
    is_path_limiting: bool,
) -> Result<gix::ObjectId> {
    // N.B. using temp index is necessary for the cases where there are conflicts in the
    // default index. I.e. by using a temp index, a subset of paths without conflicts
    // may be formed into a coherent tree while leaving the default index as-is.
    let stupid = stack.repo.stupid();
    if is_path_limiting {
        let tree_id_result = stupid.with_temp_index(|stupid_temp| {
            stupid_temp.read_tree(stack.get_branch_head().tree_id()?.detach())?;
            stupid_temp.update_index(Some(refresh_paths))?;
            stupid_temp.write_tree()
        });
        stupid.update_index(Some(refresh_paths))?;
        tree_id_result
    } else {
        if !refresh_paths.is_empty() {
            stupid.update_index(Some(refresh_paths))?;
        }
        stupid.write_tree()
    }
}

pub(crate) fn assemble_refresh_tree(
    stack: &Stack,
    matches: &ArgMatches,
    limit_to_patchname: Option<&PatchName>,
) -> Result<gix::ObjectId> {
    let stupid = stack.repo.stupid();
    let opt_pathspecs = matches.get_many::<PathBuf>("pathspecs");
    let is_path_limiting = limit_to_patchname.is_some() || opt_pathspecs.is_some();
    let statuses;

    let refresh_paths = if matches.get_flag("index") {
        // When refreshing from the index, no path limiting may be used.
        assert!(!is_path_limiting);
        IndexSet::new()
    } else {
        let maybe_patch_commit = limit_to_patchname.map(|pn| stack.get_patch_commit(pn));
        let submodules_flag = matches.get_flag("submodules");
        let nosubmodules_flag = matches.get_flag("no-submodules");
        let use_submodules = if !submodules_flag && !nosubmodules_flag {
            stack
                .repo
                .config_snapshot()
                .boolean("stgit.refreshsubmodules")
                .unwrap_or(false)
        } else {
            submodules_flag
        };
        let mut status_opts = StatusOptions::default();
        status_opts.include_submodules(use_submodules);
        if let Some(pathspecs) = opt_pathspecs {
            status_opts.pathspecs(pathspecs);
        }
        statuses = stupid.statuses(Some(&status_opts))?;

        determine_refresh_paths(
            &stupid,
            &statuses,
            maybe_patch_commit,
            matches.get_flag("force"),
        )?
    };

    let tree_id = write_tree(stack, &refresh_paths, is_path_limiting)?;

    let tree_id = if matches.get_flag("no-verify")
        || !run_pre_commit_hook(stack.repo, matches.get_flag("edit"))?
        || !stupid.diff_index_quiet(tree_id)?
    {
        tree_id
    } else {
        // Update index and rewrite tree if hook updated files in index
        write_tree(stack, &refresh_paths, is_path_limiting)?
    };

    Ok(tree_id)
}
